File: code/Java/GuessNumberGame.java
import java.util.Scanner;
import java.util.concurrent.ThreadLocalRandom;

// Inheritance
public class GuessNumberGame extends Game{
    private int minNumber;
    private int maxNumber;

    public GuessNumberGame(int minNumber, int maxNumber) {
        this.minNumber = minNumber;
        this.maxNumber = maxNumber;
    }

    @Override
    protected int generateTargetNumber() {
        return ThreadLocalRandom.current().nextInt(minNumber, maxNumber + 1);
    }

    @Override
    protected int guessNumber() {
        Scanner sn = new Scanner(System.in);
        System.out.print("\nGuess the number: ");
        return sn.nextInt();
    }
}

File: code/C/GuessNumberGame.c
#include "GuessNumberGame.h"

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int generateTargetNumber(Game* game) {
    GuessNumberGame* guessNumberGame = (GuessNumberGame*)game;
    return (rand() %
            (guessNumberGame->maxNumber - guessNumberGame->minNumber + 1)) +
           guessNumberGame->minNumber;
}

int guessNumber(Game* game) {
    int guess;
    printf("\nGuess the number: ");
    scanf("%d", &guess);
    return guess;
}

GuessNumberGame* createGuessNumberGame(int minNumber, int maxNumber) {
    GuessNumberGame* game = (GuessNumberGame*)malloc(sizeof(GuessNumberGame));
    game->game.numberOfGuesses = 1;
    game->game.totalScore = 0;
    game->game.generateTargetNumber = generateTargetNumber;
    game->game.guessNumber = guessNumber;
    game->game.startGame = startGame;
    game->game.startGame2 = startGame2;
    game->minNumber = minNumber;
    game->maxNumber = maxNumber;

    return game;
}
